File: Java/soln-heap-problems/last-stone-weight-with-max-heap.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/last-stone-weight
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> maxHeap=new PriorityQueue<>(Collections.reverseOrder());
        
        for(int stone:stones){
            maxHeap.add(stone);
        }
        
        while(maxHeap.size()>1){
            int y=maxHeap.remove();
            int x=maxHeap.remove();

            if(x!=y){
                y=y-x;
                maxHeap.add(y);
            }    
        }
        return maxHeap.isEmpty()?0:maxHeap.remove();

    }
}

File: C++/soln-heap-problems/last-stone-weight-with-max-heap.cpp
/**
 @author Farheen Bano
  
 Date 19-07-2021

 Reference-
 https://leetcode.com/problems/last-stone-weight
*/

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        priority_queue<int, vector<int>> maxHeap;
        
        for(int stone:stones){
            maxHeap.push(stone);
        }
        
        while(maxHeap.size()>1){
            int y=maxHeap.top();
            maxHeap.pop();
            int x=maxHeap.top();
            maxHeap.pop();
            if(x!=y){
                y=y-x;
                maxHeap.push(y);
            }    
   
        }
        return maxHeap.empty()?0:maxHeap.top();
    }
}; 